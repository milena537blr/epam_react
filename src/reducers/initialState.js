export default {
  movies: [],
  filters: {
    text: "",
    sortBy: "rating",
    searchBy: "title"
  }
};
