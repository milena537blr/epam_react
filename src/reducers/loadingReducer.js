import * as types from "../actions/actionTypes";
import { initialState } from "./initialState";

const loadingReducer = (state = initialState.isLoading, action) => {
  switch (action.type) {
    case types.LOADING:
      return action.isLoading;
    default:
      return state;
  }
};

export { loadingReducer };
