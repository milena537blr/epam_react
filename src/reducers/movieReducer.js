import * as types from "../actions/actionTypes";
import initialState from "./initialState";

export default function movieReducer(state = initialState.movies, action) {
  switch (action.type) {
    case types.LOAD_MOVIES_SUCCESS:
      // return action.movies;
      // return Object.assign([], state, action.movies);
      console.log(Object.assign([], state, action.movies));
    case types.ADD_TODO:
      return [
        ...state,
        {
          text: action.text,
          completed: false
        }
      ];
    default:
      return state;
  }
}
