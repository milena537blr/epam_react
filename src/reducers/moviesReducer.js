import * as types from "../actions/actionTypes";
import { initialState } from "./initialState";
import { call, put, all, takeLatest } from "redux-saga/effects";
import { updateMovies, updateCurrentMovie } from "../actions/actions";

// Sagas
export function* fetchMoviesAsync() {
  const response = yield call(
    fetch,
    "http://react-cdp-api.herokuapp.com/movies"
  );
  const movies = yield response.json();

  yield put(updateMovies(movies));
}

export function* watchFetchMovies() {
  yield takeLatest(types.FETCH_MOVIES, fetchMoviesAsync);
}

export function* fetchMovieByIdAsync(action) {
  const response = yield call(
    fetch,
    `http://react-cdp-api.herokuapp.com/movies/${action.payload}`
  );
  const movie = yield response.json();

  yield put(updateCurrentMovie(movie));
}

export function* watchFetchMovieById() {
  yield takeLatest(types.FETCH_MOVIE_BY_ID, fetchMovieByIdAsync);
}

// Movies Saga
export function* moviesSaga() {
  yield all([watchFetchMovies(), watchFetchMovieById()]);
}

const moviesReducer = (state = initialState.data, action) => {
  switch (action.type) {
    case types.LOAD_MOVIES_LOADING:
      return { ...state, loading: true, error: "" };
    case types.LOAD_MOVIES_SUCCESS:
      return { ...state, loading: false, movies: action.movies };
    case types.LOAD_MOVIES_FAILURE:
      return { ...state, loading: false, error: action.error };
    default:
      return state;
  }
};

export { moviesReducer };
