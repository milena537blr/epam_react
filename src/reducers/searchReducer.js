import * as types from "../actions/actionTypes";
import initialState from "./initialState";

export default function searchReducer(state = initialState.movies, action) {
  switch (action.type) {
    case types.SET_VISIBILITY_FILTER:
      return action.filter;
    default:
      return state;
  }
}