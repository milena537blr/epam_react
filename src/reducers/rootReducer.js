import { combineReducers } from "redux";
import { movieReducer as movies } from "./movieReducer";
import { searchReducer as filters } from "./searchReducer";

const rootReducer = combineReducers({
  movies,
  filters
});

export { rootReducer };
