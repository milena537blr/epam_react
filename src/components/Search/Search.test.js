import React from "react";
import { mount } from "enzyme";
import { createStore } from "redux";
import { Provider } from "react-redux";
import { render, fireEvent, cleanup } from "@testing-library/react";
import { initialState } from "../../reducers/initialState";
import { searchReducer } from "../../reducers/searchReducer";
import ConnectedSearch from "./Search";
import { act } from 'react-dom/test-utils';

const activeButtonClass = "red";
const inactiveButtonClass = "gray";

afterEach(cleanup);

function renderWithRedux(
  ui,
  { initialState, store = createStore(searchReducer, initialState) } = {}
) {
  return {
    ...render(<Provider store={store}>{ui}</Provider>),
    // adding `store` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    store
  };
}

test("sets up active genre button", () => {
  const { getByTestId } = renderWithRedux(<ConnectedSearch />, {
    initialState: {
      filters: {
        text: "",
        sortBy: "rating",
        searchBy: "genre"
      }
    }
  });
  expect(getByTestId("search-by-genre")).toHaveClass(activeButtonClass);
  expect(getByTestId("search-by-title")).toHaveClass(inactiveButtonClass);
});

test("sets up active title button", () => {
  const { getByTestId } = renderWithRedux(<ConnectedSearch />, {
    initialState: {
      filters: {
        text: "",
        sortBy: "rating",
        searchBy: "title"
      }
    }
  });

  expect(getByTestId("search-by-title")).toHaveClass(activeButtonClass);
  expect(getByTestId("search-by-genre")).toHaveClass(inactiveButtonClass);

  act(() => {
    fireEvent.click(getByTestId("search-by-genre"));
  });

  expect(getByTestId("search-by-title")).toHaveClass(inactiveButtonClass);
  expect(getByTestId("search-by-genre")).toHaveClass(activeButtonClass);
});

test("can render with redux with custom store", () => {
  // this is a silly store that can never be changed
  const store = createStore(() => ({
    filters: {
      text: "",
      sortBy: "rating",
      searchBy: "title"
    }
  }));
  const { getByTestId } = renderWithRedux(<ConnectedSearch />, {
    store
  });
  expect(getByTestId("search-by-title")).toHaveClass(activeButtonClass);
  expect(getByTestId("search-by-genre")).toHaveClass(inactiveButtonClass);
});

/* test("snapshot test", () => {
  const component = mount(<Search />);
  expect(component).toMatchSnapshot();
}); */
