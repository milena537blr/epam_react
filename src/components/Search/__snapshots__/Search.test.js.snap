// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="search"
        data-testid="search"
      >
        <form>
          <fieldset>
            <legend
              class="searchTitle header__searchTitle"
            >
              Find your movie
            </legend>
            <div
              class="root horizontal"
              style="margin-bottom: 24px; min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
            >
              <input
                aria-label="Search"
                aria-labelledby="searchButton"
                class="searchInput"
                name="search"
                placeholder="Quentin Tarantino"
                type="text"
              />
            </div>
            <div
              class="root horizontal spaceBetween middle"
              style="min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
            >
              <div
                class="root horizontal spaceBetween middle"
                style="min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
              >
                <div
                  class="searchLabel"
                >
                  search by
                </div>
                <div
                  class="root horizontal"
                  style="margin-right: 12px; min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
                >
                  <button
                    class="button medium red"
                    data-testid="search-by-title"
                    type="button"
                    value="title"
                  >
                    title
                  </button>
                </div>
                <div
                  class="root horizontal"
                  style="margin-right: 12px; min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
                >
                  <button
                    class="button medium gray"
                    data-testid="search-by-genre"
                    type="button"
                    value="genre"
                  >
                    genre
                  </button>
                </div>
              </div>
              <button
                class="button large red"
                type="button"
              >
                SEARCH
              </button>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="search"
      data-testid="search"
    >
      <form>
        <fieldset>
          <legend
            class="searchTitle header__searchTitle"
          >
            Find your movie
          </legend>
          <div
            class="root horizontal"
            style="margin-bottom: 24px; min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
          >
            <input
              aria-label="Search"
              aria-labelledby="searchButton"
              class="searchInput"
              name="search"
              placeholder="Quentin Tarantino"
              type="text"
            />
          </div>
          <div
            class="root horizontal spaceBetween middle"
            style="min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
          >
            <div
              class="root horizontal spaceBetween middle"
              style="min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
            >
              <div
                class="searchLabel"
              >
                search by
              </div>
              <div
                class="root horizontal"
                style="margin-right: 12px; min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
              >
                <button
                  class="button medium red"
                  data-testid="search-by-title"
                  type="button"
                  value="title"
                >
                  title
                </button>
              </div>
              <div
                class="root horizontal"
                style="margin-right: 12px; min-width: undefined; max-width: undefined; min-height: undefined; max-height: undefined;"
              >
                <button
                  class="button medium gray"
                  data-testid="search-by-genre"
                  type="button"
                  value="genre"
                >
                  genre
                </button>
              </div>
            </div>
            <button
              class="button large red"
              type="button"
            >
              SEARCH
            </button>
          </div>
        </fieldset>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
  "unmount": [Function],
}
`;
